### ISSUE TITLE
`Log::Metadata#extend` should append not prepend data (like e.g. `NamedTuple#merge`)

### ISSUE THREAD

**Issue by spuun** [crystal-lang/crystal#13360]
Similar data structures and similar methods should behave in the same way. 

`Log::Metadata#extend(other)` will put `other` first in the new `Metadata` instance, which is unexpected consider how other similar methods like `NamedTuple#merge` and `Hash#merge` works.

```crystal
a = {a: 1, b: 2}
b = {c: 3, d: 4}

a_b = a.merge(b)
puts a_b # => {a: 1, b: 2, c: 3, d: 4}

require "log"
md_a = Log::Metadata.build(a)
md_a_b = md_a.extend(b)

puts md_a_b.to_h # => {:c => 3, :d => 4, :a => 1, :b => 2}
```
I'd expect the latter to print `{:a => 1, :b => 2, :c => 3, :d => 4}`

**Comment by spuun** [comment]
Any thoughts about this? I think I can fix it.

**Comment by straight-shoota** [comment]
I'm not sure if this is easy to change due to the custom memory layout. `merge` puts the new fields first in the entry, and before iterating, `#each` calls `#defrag` which appends the parent items.
If you have an idea how this could work, go for it.
I think the chance should be accepted, it's just a matter of implementation.

@bcardiff Any thoughts on this?

**Comment by spuun** [comment]
Yeah, but I think I have a working copy locally that puts the current instance's entries last in `@first`. `#defrag` will then first copy parent entries to the beginning of `@first`, and move it's own entries to the end of parent elements or overwrite any with the same key.

**Comment by spuun** [comment]
So, the terminology here is `extend` and not `merge` so I think it's fine that it's not working exactly like `Hash#merge`, but at least I think it should append data so you can add more metadata and still get "aligned" logs.

If you "extend" with an existing key, maybe you're doing it wrong :)

**Comment by bcardiff** [comment]
I think is fine to change the `defrag` implementation to anything that will output a more expected print output.

It would also be the case that we could want a less optimized `to_s` (as in one that might not defrag) if that is easier, but it was easier to rely on `#each` for printing.

I haven't looked at the PR but from how the code is now on master maybe a way to accomplish this is to swap the entries after the parent has been defrag after the current metadata. But if another algorithm can perform that in-place directly it should be fine.
